numpy/lib/tests/test_shape_base.py

```diff
@@ -50,13 +56,50 @@ def minimal_function(array):
             apply_along_axis(minimal_function, 0, a), np.array([1, 1, 1])
         )
 
-    def test_scalar_array(self):
+    def test_scalar_array(self, cls=np.ndarray):
+        a = np.ones((6, 3)).view(cls)
+        res = apply_along_axis(np.sum, 0, a)
+        assert_(isinstance(res, cls))
+        assert_array_equal(res, np.array([6, 6, 6]).view(cls))
+
+    def test_0d_array(self, cls=np.ndarray):
+        a = np.ones((6, 3)).view(cls)
+        res = apply_along_axis(lambda x: np.squeeze(np.sum(x, keepdims=True)), 0, a)
+        assert_(isinstance(res, cls))
+        assert_array_equal(res, np.array([6, 6, 6]).view(cls))
+
+        res = apply_along_axis(lambda x: np.squeeze(np.sum(x, keepdims=True)), 1, a)
+        assert_(isinstance(res, cls))
+        assert_array_equal(res, np.array([3, 3, 3, 3, 3, 3]).view(cls))
+
+    def test_axis_insertion(self, cls=np.ndarray):
+        a = np.arange(18).reshape((6, 3))
+        res = apply_along_axis(lambda x: np.diag(x).view(cls), 0, a)
+        assert_(isinstance(res, cls))
+        assert_equal(res.ndim, 3)
+        assert_array_equal(res[:,:,0], np.diag(a[:,0]).view(cls))
```

Comment:
A cleaner way to check this sort of thing is to construct the desired result array and just call `assert_array_equal` (e.g., `np.stack([np.diag(a[:,i]) for i in range(3)], axis=-1).view(cls)`). Otherwise it's easy to overlook one part of the equality check (e.g., shape, in this case).